src/pixelator/pixeldataset/precomputed_layouts.py

```diff
@@ -0,0 +1,350 @@
+"""Pre-computed layouts for components.
+
+This module contains the functionality to generate and work with pre-computed
+graph layouts. The primary purpose of this is to allow components to be
+visualized quickly in downstream analysis, since layout computation
+is a relatively computationally expensive operation.
+
+Copyright (c) 2023 Pixelgen Technologies AB.
+"""
+
+from __future__ import annotations
+
+from multiprocessing import get_context
+from typing import TYPE_CHECKING, Iterable, Optional
+
+import pandas as pd
+import polars as pl
+
+from pixelator.graph import Graph
+from pixelator.exceptions import PixelatorBaseException
+from pixelator.marks import experimental
+from pixelator.utils import batched
+
+if TYPE_CHECKING:
+    from pixelator.pixeldataset import PixelDataset
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class PreComputedLayoutsEmpty(PixelatorBaseException):
+    """Raised when trying to access an empty PreComputedLayouts instance."""
+
+
+class PreComputedLayouts:
+    """Pre-computed layouts for a set of graphs, typically per component."""
+
+    DEFAULT_PARTITIONING = [
+        "graph_projection",
+        "layout",
+        "component",
+    ]
+
+    def __init__(
+        self,
+        layouts_lazy: pl.LazyFrame | Iterable[pl.LazyFrame] | None,
+        partitioning: Optional[list[str]] = None,
+    ) -> None:
+        """Initialize the PreComputedLayouts instance."""
+        self._layouts_lazy = layouts_lazy
+        if not partitioning:
+            partitioning = PreComputedLayouts.DEFAULT_PARTITIONING
+        self._partitioning = partitioning
+
+    @staticmethod
+    def create_empty():
+        """Create an empty PreComputedLayouts instance."""
+        return PreComputedLayouts(None)
+
+    def __repr__(self) -> str:
+        """Return a string representation of the PreComputedLayouts instance."""
+        has_data = "empty" if self.is_empty else "with data"
+        return f"PreComputedLayouts({has_data})"
+
+    def __str__(self) -> str:
+        """Return a string representation of the PreComputedLayouts instance."""
+        return self.__repr__()
+
+    @property
+    def is_empty(self) -> bool:
+        """Return True if the layout is empty."""
+        try:
+            if self.lazy is None:
+                return True
+            return self.lazy.select(pl.col("component")).first().collect().is_empty()
+        except PreComputedLayoutsEmpty:
+            return True
+
+    @property
+    def partitioning(self) -> list[str]:
+        """Return the partitioning of the layouts.
+
+        This is used to determine how to partition the data when
+        serializing it to disk, for the cases when this is applicable.
+        For example when writing hive style parquet files.
+        """
+        return self._partitioning
+
+    @property
+    def df(self) -> pd.DataFrame:
+        """Return the layouts as a pandas DataFrame."""
+        return self.lazy.collect().to_pandas()
+
```

Comment:
Small nitpick but I think this would be better renamed to a `to_pandas()` function to make it more clear that this doing some work and data is copied. Alternatively, it could be a cached_property ?

Reply:
I was inspired by this: https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.to_df.html My thinking was that since `anndata` objects has that type of function if would be good for the user to be able to do something similar here.

However, I agree it would be good to indicate that it actually does some work. Perhaps we should make it a function, and call it `to_df()`, then we would be even more similar to the `anndata` api, and we would also indicate that the user should expect that some work is being done. What do you think?

Reply:
Fully agree with making it a `to_df()` function.